vectordb_bench/backend/clients/memorydb/memorydb.py

```diff
@@ -0,0 +1,256 @@
+        else:
+            self.ef_runtime_str = ""
+        yield
+        self.conn.close()
+        self.conn = None
+
+    def ready_to_load(self) -> bool:
+        pass
+
+    def optimize(self) -> None:
+        self._post_insert()
+
+    def insert_embeddings(
+        self,
+        embeddings: list[list[float]],
+        metadata: list[int],
+        **kwargs: Any,
+    ) -> Tuple[int, Optional[Exception]]:
+        """Insert embeddings into the database.
+        Should call self.init() first.
+        """
+
+        try:
+            with self.conn.pipeline(transaction=False) as pipe:
+                for i, embedding in enumerate(embeddings):
+                    embedding = np.array(embedding).astype(np.float32)
+                    pipe.hset(metadata[i], mapping = {
+                        "id": str(metadata[i]),
+                        "metadata": metadata[i], 
+                        "vector": embedding.tobytes(),
+                    })
+                    # Execute the pipe so we don't keep too much in memory at once
+                    if (i + 1) % self.insert_batch_size == 0:
+                        pipe.execute()
+
+                pipe.execute()
+                result_len = i + 1
+        except Exception as e:
+            return 0, e
+        
+        return result_len, None
+    
+    def _post_insert(self):
+        """Wait for indexing to finish"""
+        client = self.get_client(primary=True)
+        log.info("Waiting for background indexing to finish")
+        args = (self.wait_for_no_activity, 5, "", client)
+        self.wait_until(*args)
+        if not self.db_config["cmd"]:
+            replica_clients = self.get_client(replicas=True)
+            for rc, host_name in replica_clients:
+                args = (self.wait_for_no_activity, 5, "", rc)
+                self.wait_until(*args)
+                log.debug(f"Background indexing completed in the host: {host_name}")
+                rc.close()
+    
+    def wait_until(
+        self, condition, interval=5, message="Operation took too long", *args
+    ):
+        while not condition(*args):
+            time.sleep(interval)
+    
+    def wait_for_no_activity(self, client: redis.RedisCluster | redis.Redis):
+        return (
+            client.info("search")["search_background_indexing_status"] == "NO_ACTIVITY"
+        )
+    
+    def wait_for_empty_db(self, client: redis.RedisCluster | redis.Redis):
+        return client.execute_command("DBSIZE") == 0
+    
+    def search_embedding(
+        self,
+        query: list[float],
+        k: int = 100,
+        filters: dict | None = None,
+        timeout: int | None = None,
+        **kwargs: Any,
+    ) -> (list[int]):
+        assert self.conn is not None
+        
+        query_vector = np.array(query).astype(np.float32).tobytes()
+        query_obj = Query(f"*=>[KNN {k} @vector $vec]").return_fields("id").paging(0, k).dialect(2)
```

Comment:
I'm not as familiar with Redis-compatible libs as I once was, but I see `dialect(2)` used in a bunch of places. Is there a way to make this call automatic? If not, can we have a constant (e.g. `QUERY_DIALECT`) with the value fixed in it?

Reply:
Will remove the dialect clause from the query.